routing/payment_lifecycle.go

```diff
@@ -2,346 +2,617 @@ package routing
+				p.paymentHash, err)
 
-		select {
-		case result, ok = <-resultChan:
+			routeErr, ok := err.(noRouteError)
 			if !ok {
-				return [32]byte{}, nil, htlcswitch.ErrSwitchExiting
+				return [32]byte{}, nil, err
 			}
 
-		case <-p.router.quit:
-			return [32]byte{}, nil, ErrRouterShuttingDown
-		}
+			// There is no route to try, and we have no active
+			// shards. This means that there is no way for us to
+			// send the payment, so mark it failed with no route.
+			if activeShards == 0 {
+				failureCode := routeErr.FailureReason()
+				log.Debugf("Marking payment %v permanently "+
+					"failed with no route: %v",
+					p.paymentHash, failureCode)
+
+				saveErr := p.router.cfg.Control.Fail(
+					p.paymentHash, failureCode,
+				)
+				if saveErr != nil {
+					return [32]byte{}, nil, saveErr
+				}
 
-		// In case of a payment failure, we use the error to decide
-		// whether we should retry.
-		if result.Error != nil {
-			log.Errorf("Attempt to send payment %x failed: %v",
-				p.payment.PaymentHash, result.Error)
+				continue
+			}
 
-			err = p.failAttempt(result.Error)
-			if err != nil {
+			// We still have active shards, we'll wait for an
+			// outcome to be available before retrying.
+			if err := shardHandler.waitForShard(); err != nil {
 				return [32]byte{}, nil, err
 			}
+			continue
+		}
+
+		// We found a route to try, launch a new shard.
+		attempt, outcome, err := shardHandler.launchShard(rt)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		// If we encountered a non-critical error when launching the
+		// shard, handle it.
+		if outcome.err != nil {
+			log.Warnf("Failed to launch shard %v for "+
+				"payment %v: %v", attempt.AttemptID,
+				p.paymentHash, outcome.err)
 
 			// We must inspect the error to know whether it was
 			// critical or not, to decide whether we should
 			// continue trying.
-			if err := p.handleSendError(result.Error); err != nil {
+			err := shardHandler.handleSendError(
+				attempt, outcome.err,
+			)
+			if err != nil {
 				return [32]byte{}, nil, err
 			}
 
-			// Error was handled successfully, reset the attempt to
-			// indicate we want to make a new attempt.
-			p.attempt = nil
+			// Error was handled successfully, continue to make a
+			// new attempt.
 			continue
 		}
 
-		// We successfully got a payment result back from the switch.
-		log.Debugf("Payment %x succeeded with pid=%v",
-			p.payment.PaymentHash, p.attempt.AttemptID)
+		// Now that the shard was successfully sent, launch a go
+		// routine that will handle its result when its back.
+		shardHandler.collectResultAsync(attempt)
+	}
+}
 
-		// Report success to mission control.
-		err = p.router.cfg.MissionControl.ReportPaymentSuccess(
-			p.attempt.AttemptID, &p.attempt.Route,
-		)
-		if err != nil {
-			log.Errorf("Error reporting payment success to mc: %v",
-				err)
-		}
+// shardHandler holds what is necessary to send and collect the result of
+// shards.
+type shardHandler struct {
```

Comment:
own file?

Reply:
It could, but I think we can keep it here for now, it is tightly coupled with the logic here.